### ISSUE TITLE
HandleFile class request review

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#3436]
I have created a `HandleFile` class which lets me to write handlers to receive files over WebSocket in a similar way to simple `Handle.GET`.

The file is available here: https://github.com/StarcounterSamples/KitchenSink/blob/master/src/KitchenSink/HandleFile.cs

The syntax to handle files:

```cs
HandleFile.GET("/KitchenSink/fileupload/upload", task => {
    Session.ScheduleTask(task.SessionId, (s, id) => {
        // Do you logic in here.
    });
});
```

I would like to add this class into Starcounter, otherwise each project which requires file upload would copy the class.

cc @alemoi, @warpech, @chrhol, @andwah

**Comment by warpech** [comment]
Speaking from experience, it is much better to not include super-high level APIs like this to the platform. Some reasons:

1. We will be forced to maintain it forever and ever.
2. There will always be someone dissatisfied with it and requesting new features. 

It is much better if we focus on making great `Handle.GET/POST/PUT/DELETE` experience.

That said, I think that the discussion about how to make file upload would make a great article on .io.

**Comment by warpech** [comment]
Talking about implementation, I propose to do it using middleware, just as ExpressJS does it: https://github.com/expressjs/multer

(Notice that ExpressJS, even though it is a purpose built web-framework, also doesn't have a built-in middleware for file uploads, but does it as a third party middleware).

We are soon going to have a very similar middleware API: https://github.com/Starcounter/Starcounter/issues/3251#issuecomment-190269117

Fun fact: ExpressJS also implements static file server as middleware. Why don't we go that way ultimately?

**Comment by miyconst** [comment]
>It is much better if we focus on making great Handle.GET/POST/PUT/DELETE experience.
>That said, I think that the discussion about how to make file upload would make a great article on .io.

I agree and support this approach, but I think that the article on `starcounter.io` won't be enough. We can provide it as a nuget package or just a shared dll, or a project on GitHub, at least something. It can be under authorship of one of us, and Starcounter won't be responsible to maintain it and keep compatibility.

About the middleware filters, do they work with WebSockets in Starcounter? Or you mean establish connection handler replace with a filter?

**Comment by per-samuelsson** [comment]
> Fun fact: ExpressJS also implements static file server as middleware. Why don't we go that way ultimately?

We will for sure, when we reach DNX, which does the same. The initiative started with middleware is to go step by step, not to just rewrite everything yet another time when users start to get used to it.

**Comment by per-samuelsson** [comment]
> We will for sure, when we reach DNX, which does the same

Here's where we are going: [Serving static files](http://docs.asp.net/en/latest/fundamentals/static-files.html#serving-static-files). Notice how you need to enable it as middleware.

**Comment by mtornwall** [comment]
~~~
this.FilePath = Path.GetTempFileName();
~~~

`Path.GetTempFileName` comes with a little caveat. From [the MSDN article](https://msdn.microsoft.com/en-us/library/system.io.path.gettempfilename(v=vs.110).aspx):

> The GetTempFileName method will raise an IOException if it is used to create more than 65535 files without deleting previous temporary files.

So it's imperative that the temporary files be cleaned up at some point. Personally I'd rather have the caller specify a temporary directory and just have a custom temporary file name generator.

**Comment by miyconst** [comment]
>`Path.GetTempFileName` comes with a little caveat.

I know, I know :). As @warpech said, it is not going to be included into Starcounter itself, instead it will be an add-on package. So, users will be able to modify it for their needs.

**Comment by miyconst** [comment]
Thank you all for the comments and reviews. I will write a blog post about it (eventually).
